sql/transform/node.go

```diff
@@ -0,0 +1,395 @@
+// Copyright 2020-2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform
+
+import (
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+// NodeFunc is a function that given a node will return that node
+// as is or transformed, a boolean to indicate whether the node was modified,
+// and an error, if any.
+type NodeFunc func(sql.Node) (sql.Node, TreeIdentity, error)
+
+// ExprFunc is a function that given an expression will return that
+// expression as is or transformed, a boolean to indicate whether the expression
+// was modified, and an error, if any.
+type ExprFunc func(sql.Expression) (sql.Expression, TreeIdentity, error)
+
+// Context is the parameter to the Transform{,Selector}.
+type Context struct {
+	// Node is the currently visited node which will be transformed.
+	Node sql.Node
+	// Parent is the current parent of the transforming node.
+	Parent sql.Node
+	// ChildNum is the index of Node in Parent.Children().
+	ChildNum int
+	// SchemaPrefix is the concatenation of the Parent's SchemaPrefix with
+	// child.Schema() for all child with an index < ChildNum in
+	// Parent.Children(). For many Node, this represents the schema of the
+	// |row| parameter that is going to be passed to this node by its
+	// parent in a RowIter() call. This field is only non-nil if the entire
+	// in-order traversal of the tree up to this point is Resolved().
+	SchemaPrefix sql.Schema
+}
+
+// CtxFunc is a function which will return new sql.Node values for a given
+// Context.
+type CtxFunc func(Context) (sql.Node, TreeIdentity, error)
+
+// Selector is a function which will allow NodeWithCtx to not
+// traverse past a certain Context. If this function returns |false|
+// for a given Context, the subtree is not transformed and the child
+// is kept in its existing place in the parent as-is.
+type Selector func(Context) bool
+
+// ExprWithNodeFunc is a function that given an expression and the node that contains it, will return that
+// expression as is or transformed along with an error, if any.
+type ExprWithNodeFunc func(sql.Node, sql.Expression) (sql.Expression, TreeIdentity, error)
+
+// TreeIdentity tracks modifications to node and expression trees
```

Comment:
Give a little more info here. When should implementors return one versus the other value?